Graph type builder + hierarchy cleanup

d-michail opened the pull request:
Added a graph builder (using the builder pattern) in order to be able to create graphs without remembering the graph class hierarchy. Called it `GraphTypeBuilder` since we already have a `GraphBuilder`.

This is a long overdue change which allows someone to create a graph like: 
```
Graph<Integer, DefaultEdge> graph = 
   GraphTypeBuilder.<Integer, DefaultEdge> directed()
.allowingMultipleEdges(true).allowingSelfLoops(true).edgeClass(DefaultEdge.class).build();
```

You can also create a graph which has the same type as another graph, like: 

```
Graph<Integer, DefaultEdge> graph = new Pseudograph<>(DefaultEdge.class);
Graph<Integer, DefaultEdge> newGraph = GraphTypeBuilder.asGraph(graph).build();
```

I also did some minor changes in the class hierarchy which we forgot to do when we changed the graph interfaces. Now that we do not have any interfaces called "DirectedGraph" or "UndirectedGraph", we can remove the "Default" prefix from the graph classes. So `DefaultDirectedGraph` is now `DirectedGraph`. I also added an `UndirectedGraph` class to complete the hierarchy.

I also cleaned up a few minor issues that Eclipse was complaining about.

jbduncan commented:
As someone who's worked with Guava's `com.google.common.graph.*` types, where there exists a very similar graph builder class, I very much love the idea of this PR. :)

jkinable commented:
Could you add the examples you state in the intro of this PR to the class description of GraphTypeBuilder? Having such examples directly in the javadoc makes it super easy to use the class (see for instance the javadoc in the standard String class).

d-michail commented:
Nice suggestion, I added some examples of use. 

Other changes: 
 - I changed the method names as suggested by @jsichi 
 - I also added the term "parallel" in the javadoc comments where we mention multiple edges.

d-michail commented:
@jsichi I took back the removal of the "Default" prefix for the graph implementations. I noted in the Javadoc that this is the default implementation of a directed graph that the library provides. I also added the undirected variants of such classes, for example "DefaultUndirectedGraph".

I also took some time improving a little bit the Javadocs inside the graph package.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/graph/builder/GraphTypeBuilder.java:
```diff
@@ -0,0 +1,341 @@
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    public static <V, E> GraphTypeBuilder<V, E> asGraphType(GraphType type)
```
The method naming convention "asX" usually pairs with a conversion to type X.  So I suggest "forGraphType" and "forGraph" instead.

jsichi commented:
Thanks!
